### PR TITLE
Updating collateral verification API to fetch custom data.

### PR DESCRIPTION
**Pull request by msft-gumunjal**
To fetch a certain tcb info for attestation, this PR exposes a new interface to fetch tcb evaluation number and other custom query parameters.
Added a test to validate the code path.

### CODE REVIEW

File: src/UnitTest/test_quote_prov.cpp [hunk 1]
```diff
@@ -509,6 +526,17 @@ static void GetVerificationCollateralTest()
         &collateral);
     ASSERT_TRUE(SGX_QL_SUCCESS == result);
     VerifyCollateral(collateral);
+
+    collateral = nullptr;
+    result = sgx_ql_get_quote_verification_collateral_with_params(
```
Comment (Shivr-ctrl) [inline src/UnitTest/test_quote_prov.cpp:537]: Instead of adding it part of the same test, could make a separate unit test, which calls collateral with params for both V2 and V3 collateral.

File: src/UnitTest/test_quote_prov.cpp [hunk 2]
```diff
@@ -509,6 +526,17 @@ static void GetVerificationCollateralTest()
...
+            TEST_FMSPC,
+            sizeof(TEST_FMSPC),
+            "processor",
+            custom_param,
+            custom_param_length,
+            &collateral);
+    ASSERT_TRUE(SGX_QL_SUCCESS == result);
+    VerifyCollateral(collateral);
```
Comment (Shivr-ctrl) [inline src/UnitTest/test_quote_prov.cpp:545]: Additional verification will be needed. Would want to confirm that the collateral received has the same tcbevaluation number as what was requested.

File: src/UnitTest/test_quote_prov.cpp [hunk 3]
```diff
@@ -616,8 +655,7 @@ boolean RunQuoteProviderTests(bool caching_enabled = false)
     local_cache_clear();
     auto duration_curl_cert = MeasureFunction(GetCertsTest);
     GetCrlTest();
-    auto duration_curl_verification =
-        MeasureFunction(GetVerificationCollateralTest);
+    auto duration_curl_verification = MeasureFunction(GetVerificationCollateralTest);
```
Comment (Shivr-ctrl) [inline src/UnitTest/test_quote_prov.cpp]: Nit: Seems like code formatting unrelated to your change

File: src/dcap_provider.cpp [hunk 4]
```diff
@@ -311,16 +313,16 @@ bool get_cert_cache_expiration_time(const string& cache_max_age, const string& u
     catch (const std::invalid_argument& e)
     {
         log(SGX_QL_LOG_ERROR,
-            "Invalid argument thrown when parsing cache-control. Header text: '%s' Error: '%s'. Collateral will not be cached",
+            "Invalid argument thrown when parsing cache-control. Header "
```
Comment (Shivr-ctrl) [inline src/dcap_provider.cpp]: nit: Seems unrelated.

File: src/dcap_provider.cpp [hunk 5]
```diff
@@ -898,19 +900,79 @@ static sgx_plat_error_t build_pck_crl_url(
     return SGX_PLAT_ERROR_OK;
 }
 
-static std::string build_tcb_info_url(const std::string& fmspc)
+char* base64_encode(const void* source)
+{
+    // Character set of base64 encoding scheme
+    char char_set[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
+    const char* input = static_cast<const char*>(source);
+    int len_str = strlen(input);
```
Comment (Shivr-ctrl) [inline src/dcap_provider.cpp:908]: This is not correct. Input is a void pointer and there is an implied assumption in the code that it is string. The param_length parameter should be input to this function and that should be used as length.

File: src/dcap_provider.cpp [hunk 6]
```diff
@@ -898,19 +900,79 @@ static sgx_plat_error_t build_pck_crl_url(
...
 }
 
-static std::string build_tcb_info_url(const std::string& fmspc)
+char* base64_encode(const void* source)
+{
+    // Character set of base64 encoding scheme
+    char char_set[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
+    const char* input = static_cast<const char*>(source);
+    int len_str = strlen(input);
+    char* res_str = (char*)malloc(SIZE * sizeof(char));
```
Comment (Shivr-ctrl) [inline src/dcap_provider.cpp:909]: This is hardcoded to 1000 any reason for that? Can this be based on size of the input?

Comment (Shivr-ctrl) [inline src/dcap_provider.cpp:909]: Where will this memory be freed?
A better interface is the caller allocates memory and passes as input and it owns freeing it after the http call is completed

File: src/dcap_provider.cpp [hunk 7]
```diff
@@ -898,19 +900,79 @@ static sgx_plat_error_t build_pck_crl_url(
...
+
+    int index, no_of_bits = 0, padding = 0, val = 0, count = 0, temp;
+    int i, j, k = 0;
+
+    for (i = 0; i < len_str; i += 3)
```
Comment (Shivr-ctrl) [inline src/dcap_provider.cpp:914]: Some comments will help. Example why += 3?

File: src/dcap_provider.cpp [hunk 8]
```diff
@@ -898,19 +900,79 @@ static sgx_plat_error_t build_pck_crl_url(
...
     }
-    tcb_info_url << "/tcb/";
-    tcb_info_url << format_as_hex_string(fmspc.c_str(), fmspc.size()) << "?";
+    tcb_info_url << "/tcb?";
+    tcb_info_url << "fmspc=" << format_as_hex_string(fmspc.c_str(), fmspc.size()) << "&";
+
+    if (custom_param != nullptr)
+    {
+        additional_parameters = base64_encode(custom_param);
+        tcb_info_url << "customParameter=" << additional_parameters << "&";
```
Comment (Shivr-ctrl) [inline src/dcap_provider.cpp:974]: Same needs to be done for QEID and QVEID also

File: src/UnitTest/test_quote_prov.cpp [hunk 9]
```diff
@@ -93,89 +101,90 @@ static sgx_ql_free_qve_identity_t sgx_ql_free_qve_identity;
...
+static sgx_ql_get_quote_verification_collateral_with_params_t
+    sgx_ql_get_quote_verification_collateral_with_params;
 static sgx_ql_get_qve_identity_t sgx_ql_get_qve_identity;
 static sgx_ql_get_root_ca_crl_t sgx_ql_get_root_ca_crl;
 
 // Test FMSPC
 static constexpr uint8_t TEST_FMSPC[] = {0x00, 0x90, 0x6E, 0xA1, 0x00, 0x00};
 static constexpr uint8_t ICX_TEST_FMSPC[] = {0x00, 0x60, 0x6a, 0x00, 0x00, 0x00};
 
+const uint16_t custom_param_length = 2;
```
[the remark this answers is not in the document]
Comment (anche-is-andy) [reply src/UnitTest/test_quote_prov.cpp:113]: yes, should be strlen(custom_param) + 1 in this case.

File: src/dcap_provider.cpp (hunk 5 above)
Comment (anche-is-andy) [reply src/dcap_provider.cpp:908]: Please don't assume the input is a string

File: src/dcap_provider.cpp [hunk 10]
```diff
@@ -898,19 +900,79 @@ static sgx_plat_error_t build_pck_crl_url(
     return SGX_PLAT_ERROR_OK;
 }
 
-static std::string build_tcb_info_url(const std::string& fmspc)
+char* base64_encode(const void* source)
```
Comment (anche-is-andy) [inline src/dcap_provider.cpp:903]: it's better to return std::string

File: src/dcap_provider.cpp [hunk 11]
```diff
@@ -898,19 +900,79 @@ static sgx_plat_error_t build_pck_crl_url(
     return SGX_PLAT_ERROR_OK;
 }
 
-static std::string build_tcb_info_url(const std::string& fmspc)
+char* base64_encode(const void* source)
+{
+    // Character set of base64 encoding scheme
+    char char_set[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
```
Comment (anche-is-andy) [inline src/dcap_provider.cpp:906]: const please

File: src/dcap_provider.cpp (hunk 6 above)
Comment (anche-is-andy) [reply src/dcap_provider.cpp:909]: std::string result instead of res_str?

File: src/dcap_provider.cpp [hunk 12]
```diff
@@ -898,19 +900,79 @@ static sgx_plat_error_t build_pck_crl_url(
...
-static std::string build_tcb_info_url(const std::string& fmspc)
+char* base64_encode(const void* source)
+{
+    // Character set of base64 encoding scheme
+    char char_set[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
+    const char* input = static_cast<const char*>(source);
+    int len_str = strlen(input);
+    char* res_str = (char*)malloc(SIZE * sizeof(char));
+
+    int index, no_of_bits = 0, padding = 0, val = 0, count = 0, temp;
```
Comment (anche-is-andy) [inline src/dcap_provider.cpp:911]: number_of_bits?

File: src/dcap_provider.cpp [hunk 13]
```diff
@@ -1699,11 +1780,13 @@ extern "C" quote3_error_t sgx_ql_free_root_ca_crl(char* p_root_ca_crl)
     return SGX_QL_SUCCESS;
 }
 
-extern "C" quote3_error_t sgx_ql_get_quote_verification_collateral(
+quote3_error_t sgx_ql_fetch_quote_verification_collateral(
```
[the remark this answers is not in the document]
Comment (anche-is-andy) [reply src/dcap_provider.cpp:1763]: If this is the old API, please don't change it, you should have the old API call _with_params new API instead of the other way around

File: src/UnitTest/test_quote_prov.cpp [hunk 14]
```diff
@@ -333,6 +359,32 @@ static HINSTANCE LoadFunctions()
 }
 #endif
 
+//
+// extract raw value from response body, if exists
+//
+sgx_plat_error_t extract_from_json(
```
Comment (Shivr-ctrl) [inline src/UnitTest/test_quote_prov.cpp:365]: Please consider using a different name than in prod code. extract_from_nested_json ?

File: src/UnitTest/test_quote_prov.cpp [hunk 15]
```diff
@@ -526,6 +601,30 @@ static void GetVerificationCollateralTestICXV3()
...
+            custom_param,
+            custom_param_length, 
+            &collateral);
+    ASSERT_TRUE(SGX_QL_SUCCESS == result);
+    nlohmann::json json_body = nlohmann::json::parse(collateral->tcb_info);
+    extract_from_json(
+        json_body.flatten(),
+        "/tcbInfo/tcbEvaluationDataNumber",
+        &tcbEvaluationNumber);
+    ASSERT_TRUE(tcbEvaluationNumber == tcbEvaluationDataNumber);
```
Comment (Shivr-ctrl) [inline src/UnitTest/test_quote_prov.cpp:624]: Should this be string compare? as both the variables are strings

File: src/Windows/dll/dcap_provider.def [hunk 16]
```diff
@@ -9,7 +9,8 @@ EXPORTS
     sgx_ql_free_qve_identity;
     sgx_ql_free_root_ca_crl;
     sgx_ql_get_quote_verification_collateral;
+    sgx_ql_get_quote_verification_collateral_with_params;
     sgx_ql_get_qve_identity;
     sgx_ql_get_root_ca_crl;
     sgx_get_qe_identity_info
-    sgx_free_qe_identity_info
```
Comment (Shivr-ctrl) [inline src/Windows/dll/dcap_provider.def:15]: Will this newline in .def file make a difference? Why changed?

File: src/dcap_provider.cpp [hunk 17]
```diff
@@ -962,6 +1106,21 @@ static std::string build_enclave_id_url(
...
+    {
+        std::string encoded_str;
+        try
+        {
+            encoded_str = base64_encode(custom_param, custom_param_length);
+        }
+        catch (exception e)
+        {
+            log(SGX_QL_LOG_ERROR, "Enclave_Id_URL: Invalid parameters provided.");
+            throw e;
```
Comment (Shivr-ctrl) [inline src/dcap_provider.cpp:1119]: Why not return SGX_PLAT_ERROR_INVALID_PARAMETER instead of throwing from here?

**Review by Shivr-ctrl — APPROVED** [review]
